File: leetcode/medium/2359leetcode.cpp
Find Closest Node to Given Two Nodes

class Solution {
public:
    void dfs(int cur,vector<int>& edges,vector<bool>&visited,vector<int>&dist){
        visited[cur]=true;
        int x=edges[cur];
        if(x!=-1 && !visited[x]){
            visited[x]=true;
            dist[x]=1+dist[cur];
            dfs(x,edges,visited,dist);
        }
        return;
    }
    int closestMeetingNode(vector<int>& edges, int node1, int node2) {
        int n=edges.size();
        vector<bool>visited(n,false);
        vector<int>dist1(n,INT_MAX),dist2(n,INT_MAX);
        dist1[node1]=0;
        dist2[node2]=0;
        dfs(node1,edges,visited,dist1);
        vector<bool>v(n,false);
        dfs(node2,edges,v,dist2);
        int maxi=INT_MAX;
        int node=-1;
        for(int i=0;i<n;i++){
            int maxd=max(dist1[i],dist2[i]);
            if(maxi>maxd){
                maxi=maxd;
                node=i;
            }
        }
        return  node;
    }
};

File: leetcode/medium/2359leetcode.java
//2359. Find Closest Node to Given Two Nodes

class Solution {
    private int n;
    private List<List<Integer>> adj = new ArrayList<>();
    public int closestMeetingNode(int[] edges, int node1, int node2) {
        n = edges.length;
        int[] dis1 = new int[n];
        int[] dis2 = new int[n];
        Arrays.fill(dis1, Integer.MAX_VALUE);
        Arrays.fill(dis2, Integer.MAX_VALUE);

        boolean[] vis = new boolean[n];

        for(int i = 0; i < n; i++){
            adj.add(new ArrayList<>());
        }

        for(int i = 0; i < n; i++){
            if(edges[i] == -1) adj.get(i).add(i);
            else adj.get(i).add(edges[i]);
        }

        dfs(node1, dis1, 0, vis);
        Arrays.fill(vis, false);
        dfs(node2, dis2, 0, vis);

        int ans = Integer.MAX_VALUE;
        int idx = -1;
        for(int i = 0; i < n; i++){
            if(ans > Math.max(dis1[i], dis2[i])){
                ans = Math.max(dis1[i], dis2[i]);
                idx = i;
            }
        }
        return idx;
    }
    private void dfs(int i, int[] dis, int d, boolean[] vis){
        vis[i] = true;
        dis[i] = d;
        for(int it : adj.get(i)){
            if(!vis[it]) dfs(it, dis, d+1, vis);
        }
    }
}